core/src/main/java/com/linecorp/armeria/common/ConnectionPoolListener.java

```diff
@@ -0,0 +1,173 @@
+/*
+ * Copyright 2024 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package com.linecorp.armeria.common;
+
+import java.net.InetSocketAddress;
+
+import com.linecorp.armeria.common.annotation.UnstableApi;
+import com.linecorp.armeria.common.metric.MeterIdPrefix;
+
+import io.micrometer.core.instrument.MeterRegistry;
+import io.netty.util.AttributeMap;
+
+/**
+ * Listens to the client connection pool events.
+ */
+public interface ConnectionPoolListener extends com.linecorp.armeria.client.ConnectionPoolListener {
```

Comment:
On second thought, `ConnectionEventListener` would be a better name for the new API than `ConnectionPoolListener`. Because:
  - If we use the same name, setter methods can be easily overloaded but it is difficult to name the new API while deprecating the old one.
  - `pool` is not a good word for the server side. Servers accept incoming connections but do not manage pooled connections for reuse. Therefore, a more general term, `event` would make more sense.  
```suggestion
public interface ConnectionEventListener {
```

Should we deprecate `ClientFactoryBuilder.connectionPoolListener()` and related methods in favor of `ClientFactoryBuilder.connectionEventListener()`?
For backward compatibility, we need to convert `ConnectionPoolListener` to `ConnectionEventListener` internally.

Reply:
I like the name. 😊

Reply:
@ikhoon

> Should we deprecate ClientFactoryBuilder.connectionPoolListener() and related methods in favor of ClientFactoryBuilder.connectionEventListener()?

Yes, I think so.

> For backward compatibility, we need to convert ConnectionPoolListener to ConnectionEventListener internally.

Okay. I'll convert it.👍

Reply:
I converted the deprecated ConnectionPoolListener into ClientConnectionEventListener.

```
// ClientConnectionEventListener.class
    static final ClientConnectionEventListenerAdapter NOOP = new ClientConnectionEventListenerAdapter();

    /**
     * Convert the {@link ConnectionPoolListener} into a {@link ClientConnectionEventListener}.
     */
    static ClientConnectionEventListener of(ConnectionPoolListener connectionPoolListener) {
        return new ClientConnectionEventListenerAdapter() {
            @Override
            public void connectionOpened(@Nullable SessionProtocol desiredProtocol,
                                         SessionProtocol protocol,
                                         InetSocketAddress remoteAddress,
                                         InetSocketAddress localAddress,
                                         AttributeMap attrs) throws Exception {
                connectionPoolListener.connectionOpen(protocol,
                                                      remoteAddress,
                                                      localAddress,
                                                      attrs);
            }

            @Override
            public void connectionClosed(SessionProtocol protocol,
                                         InetSocketAddress remoteAddress,
                                         InetSocketAddress localAddress,
                                         boolean isIdle,
                                         AttributeMap attrs) throws Exception {
                connectionPoolListener.connectionClosed(protocol,
                                                        remoteAddress,
                                                        localAddress,
                                                        attrs);
            }
        };
    }
```
```
// ClientFactoryBuilder.class

    private ClientFactoryOptions buildOptions() {
       ...
        if (options.containsKey(ClientFactoryOptions.CONNECTION_POOL_LISTENER) &&
            !options.containsKey(ClientFactoryOptions.CONNECTION_EVENT_LISTENER)) {
            final ConnectionPoolListener connectionPoolListener =
                    (ConnectionPoolListener) options.get(ClientFactoryOptions.CONNECTION_POOL_LISTENER).value();

            options.put(ClientFactoryOptions.CONNECTION_EVENT_LISTENER,
                        ClientFactoryOptions.CONNECTION_EVENT_LISTENER
                                .newValue(ClientConnectionEventListenerAdapter.of(connectionPoolListener)));
        }
```